Support for Azure consumption pricesheet download API

babbageclunk opened the pull request:
## What does this PR change?
* It extends the Azure provider to get customer-specific prices using the consumption pricesheet download API, if the user provides a billing account and the configured credentials have access to request it.

## Does this PR relate to any other PRs?
* Not that I'm aware of.

## How will this PR impact users?
* If their cluster is running on Azure and they have special prices negotiated, OpenCost will give more accurate pricing.

## Does this PR address any GitHub or Zendesk issues?
* No

## How was this PR tested?
* By running the custom image in a cluster with a billing account added to the `azure-service-key` secret, checking the logs and the `/pricingSourceStatus`, `/allNodePricing` API endpoints.
* Also by running the custom image without a billing account to ensure that the current ratecard handling still works.

## Does this PR require changes to documentation?
* Yes, I'll need to add information on specifying the billing account in the `azure-service-key` - at the moment the only documentation for the `azure-service-key` is on the Kubecost side. It also needs details about how to grant the service principal access to the billing account.

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next Opencost release? If not, why not?
* I'm not sure whether this should be in the next release - it's intended for the kubecon candidate.

babbageclunk committed:
Initial support for Azure consumption pricesheet download API

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

CsBigDataHub commented:
Is the related to https://github.com/opencost/opencost/issues/1757 ?

ameijer commented on pkg/cloud/azurepricesheet/client.go:
```diff
@@ -0,0 +1,126 @@
+		return nil, err
+	}
+	reqQP := req.Raw().URL.Query()
+	reqQP.Set("api-version", "2022-06-01")
```
does this potentially conflict with the comment on line 86 re: a 2021-10-01 API?

babbageclunk replied:
Oops, like this!

ameijer commented on pkg/cloud/azurepricesheet/client.go:
```diff
@@ -0,0 +1,126 @@
+
+// getByBillingPeriodCreateRequest creates the GetByBillingPeriod request.
+func (client *PriceSheetClient) downloadByBillingPeriodCreateRequest(ctx context.Context, billingPeriodName string) (*policy.Request, error) {
+	urlPath := "/providers/Microsoft.Billing/billingAccounts/{billingAccountID}/billingPeriods/{billingPeriodName}/providers/Microsoft.Consumption/pricesheets/download"
```
thoughts on making this a const then sprintf-ing both vars at the same time to render the final URL?

babbageclunk replied:
Good call - I'll do that. This was adapted pretty quickly from existing azure-sdk-for-go code, I'll do a pass over it to make sure there aren't any inconsistencies.

ameijer commented on pkg/cloud/azureprovider.go:
```diff
@@ -847,127 +858,378 @@ func (az *Azure) DownloadPricingData() error {
+		pricings, err := convertMeterToPricings(v, regions, baseCPUPrice)
 		if err != nil {
+			log.Warnf("converting meter to pricings: %s", err.Error())
 			continue
```
do you think an error in this method is serious enough to return it immediately?

babbageclunk replied:
I was keeping this behaviour the same as it is for the ratecard handling code for consistency, but I'm alright making this an error that stops the pricesheet processing if you'd prefer.

ameijer replied:
your call, sounds like you considered the tradeoffs already. Just wanted to have a discussion!

ameijer commented on pkg/cloud/azureprovider.go:
```diff
@@ -847,127 +858,378 @@ func (az *Azure) DownloadPricingData() error {
-	az.RateCardPricingError = nil
+}
+
+type pricesheetDownloader struct {
```
Perhaps I am misunderstanding, but what are your thoughts on placing this in the price sheet client file you developed? From my reading, the current purpose of the price sheet client is to obtain a download URL for the pricesheet, then this pricesheet downloader downloads the actual pricesheet itself. Would we do one operation without the other? If not, maybe that can just all be abstracted into one call to the client.

babbageclunk replied:
I did think about this but ended up not doing it because of the dependency it has on `convertMeterToPricings`, which is also used from the ratecard code. 

But now that I think about it more I could pass that function in as config. I'll do that, I think it makes sense to keep the cloud package tidier since all the providers are in it together.

ameijer commented on pkg/cloud/azureprovider.go:
```diff
@@ -847,127 +858,378 @@ func (az *Azure) DownloadPricingData() error {
+	return out, nil
+}
+
+func (d *pricesheetDownloader) readPricesheet(ctx context.Context, data io.Reader) (map[string]*AzurePricing, error) {
```
@babbageclunk do you think this would be a good place for a unit test? What we have been trying to do is feed in a few lines of cloud provider pricing data into these response body readers (e.g., https://github.com/opencost/opencost/blob/develop/pkg/cloud/gcpprovider_test.go#L124) just to test the happy path.

babbageclunk replied:
Yeah you're right, this is totally amenable to some unit tests, I'll add them here.

ameijer reviewed (commented):
hello @babbageclunk! This looks awesome - a big step forward for this component. Took a quick pass. Would be interested to hear your perspective on getting some testing on this file in

babbageclunk committed:
Match handling of unknown regions in previous version

These are silently skipped - logging them was very alarming!

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

babbageclunk committed:
Add some logging in the pricesheet download

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

babbageclunk committed:
Skip the first two lines of the pricesheet

They don't have commas, so we can't skip them with the CSV reader
because the field count won't match. Skip them using the buffered
reader first.

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

babbageclunk committed:
Temporary changes while testing

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

ameijer commented:
@babbageclunk happy to give additional reviews as you would like, just re-request one from me when ready!

babbageclunk committed:
Move pricesheetDownloader to azurepricesheet.Downloader

This avoids polluting the cloud package.

Pass a function to convert meter info into pricings in as config for
the downloader so we can share the convertMeterToPricings function
between the rate card and price sheet paths.

Similarly, since AzurePricing is defined in the cloud package, make
Downloader generic over that type. I'm not sure about this, it seems a
little clever - is there a sensible place to move AzurePricing to so
it can be imported in both pkg/cloud and pkg/cloud/azurepricesheet?

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

babbageclunk committed:
Add Azure File pricing in the pricesheet path

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

babbageclunk commented on pkg/cloud/azurepricesheet/downloader.go:
```diff
@@ -0,0 +1,252 @@
+	"github.com/opencost/opencost/pkg/log"
+)
+
+type Downloader[T any] struct {
```
I made this generic over the specific pricing type to avoid a circular dependency on `pkg/cloud.AzurePricing` in this code, but that might be too clever. Is there a sensible location I could move `AzurePricing` to so it can be imported from `pkg/cloud` and `pkg/cloud/azurepricesheet`?

babbageclunk replied:
I guess I could just move `AzurePricing` into `pkg/cloud/azurepricesheet`

ameijer replied:
@babbageclunk what if you moved all of this + the existing azureProvider into a new package pkg/cloud/azure? You have made the decision (that I concur with) to introduce a new azure specific subpackage (azurepricesheet) - it could be refactored to just azure and include the azureprovider.go to become a one stop shop for all azure specific logic. 

This would almost certainly break our enterprise builds downstream but I am happy to prepare a companion PR and take care of things here on our end. 

That seems more maintainable than a pseudo-generic - what do you think?

babbageclunk replied:
That sounds like a great idea! @r2k1 had mentioned that there was a reorganisation happening in parallel to break the providers into separate packages. I'll make a quick PR to do that and remove the type parameter.

babbageclunk committed:
Get Azure billing account and offer ID from environment

They're not sensitive information so it's much simpler than passing
them in through the service key secret.

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

babbageclunk committed:
Rename Downloader.Run to .GetPricing

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

babbageclunk committed:
Tidy up the pricesheet client code based on review

Still had some inconsistencies from being adapted from a different
synchronous API.

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

babbageclunk committed:
Tweak logging to clarify what region is unsupported

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

babbageclunk committed:
Add price sheet API information to the /pricingSourceStatus endpoint

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

AjayTripathy commented:
Hey @babbageclunk did you want any help producing docs? and @ameijer are we ready to take another pass?

babbageclunk committed:
Normalise units in the price sheet

The records in the price sheet have a variety of units, while the code
in the azure provider that extracts prices expects the units to be "1
/Hour" for VMs or "1 /Month" for storage (which it then converts to
GB/Hour).

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

ameijer commented:
@AjayTripathy I was waiting for @babbageclunk 's ping that he had addressed the items from the last look before I take another pass. Am monitoring this closely though 👍

babbageclunk commented:
Just working on the unit test for the downloader now. 

My plan for docs is to pull [this page](https://docs.kubecost.com/install-and-configure/install/cloud-integration/azure-out-of-cluster/azure-config) across from the Kubecost documentation, adapt it to fit OpenCost, and then add information about the pricesheet configuration to it:
1. assigning the billing access role to the SP
2. passing the offer ID and billing account in environment variables

For 1, do you think it would be better to include the bash script to assign the role, or to spell out the steps? The steps are a bit messy since there isn't Azure CLI support for assigning a billing role to an SP as far as I can tell - you need to do a PUT to an ARM URL directly.

babbageclunk commented:
Actually, it would be helpful if someone there could migrate the Kubecost rate card documentation to the OpenCost docs?

babbageclunk committed:
Unit tests for Downloader.readPricesheet

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

babbageclunk committed:
Reformatted Downloader tests

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

babbageclunk committed:
Add basic tests for convertMeterToPricings

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

babbageclunk committed:
Remove the price sheet temp file on close

Signed-off-by: Christian Muirhead <christian.muirhead@microsoft.com>

ameijer reviewed (approved):
@babbageclunk this lgtm - tests look good as well. Perhaps consider merging this, then doing that refactor of azurepricing in a follow up PR since it is not a functionality blocker?

AjayTripathy commented:
Thanks so much @babbageclunk ! Anywhere I can get started with helping to write documentation for how to set up the key on the Azure side?
